### DISCUSSION TITLE
AYUDA EN ERROR DE MEMORIA EN LEAVE PENGUIN

### DISCUSSION THREAD

**Discussion by SantiagoOsorio3** [IIC2133-PUC/2024-2#41] [Tarea 0]
Hola, llevo desde el jueves con un error de memoria en la función leave penguin el cual NO puedo solucionar. Quería preguntar si algún ayudante me podria ayudar porfavor....

(se que mi error es que estoy accediendo a partes de memoria que ya fueron eliminadas pero no he sido capaz de debbugearlo).

**Comment by 4n1t4** [comment]
Hola 👋,  manda lo que sale en valgrind y así puedo ver bien tu error. No envíes código, ese lo veo en tu repo.

**Reply by SantiagoOsorio3** [reply]
bakan, acabo de hacer un push de lo que llevaba, pero nose si estaba bien.... pero revisemos de ahi.

esto me arroja valgrind 

root@9806622cfb78:~/volume/probar_setup# valgrind --leak-check=full --show-leak-kinds=all ./program input.txt output.txt
==794== Memcheck, a memory error detector
==794== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==794== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==794== Command: ./program input.txt output.txt
==794== 
ENTER
Penguin 1 with 3 Puffles:
blue
green
black
ENTER
Penguin 3 with 1 Puffles:
blue
ENTER
Entry denied due to maximum capacity
STATUS
STATUS:
	Total penguins: 2
	Penguin 1:
	Penguin 3:
	Total puffles: 4
BUY-PUFFLE
 1 1
Purchase denied
 3 1
BUY-PUFFLE
 1 3
 3 3
Penguin 3 bought 9
TRADE-PUFFLE
Exchange between 1 and 3
STEAL-PUFFLE
Steal 1 from 3
	Total stolen puffles: 0
LEAVE
==794== Invalid read of size 8
==794==    at 0x109184: leave_penguin (in /root/volume/probar_setup/program)
==794==    by 0x10A1DF: main (in /root/volume/probar_setup/program)
==794==  Address 0x49da4b8 is 8 bytes before a block of size 16 alloc'd
==794==    at 0x484C0A4: calloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x109E87: main (in /root/volume/probar_setup/program)
==794== 
==794== Invalid read of size 8
==794==    at 0x109188: leave_penguin (in /root/volume/probar_setup/program)
==794==    by 0x10A1DF: main (in /root/volume/probar_setup/program)
==794==  Address 0x8 is not stack'd, malloc'd or (recently) free'd
==794== 
==794== 
==794== Process terminating with default action of signal 11 (SIGSEGV)
==794==  Access not within mapped region at address 0x8
==794==    at 0x109188: leave_penguin (in /root/volume/probar_setup/program)
==794==    by 0x10A1DF: main (in /root/volume/probar_setup/program)
==794==  If you believe this happened as a result of a stack
==794==  overflow in your program's main thread (unlikely but
==794==  possible), you can try to increase the size of the
==794==  main thread stack using the --main-stacksize= flag.
==794==  The main thread stack size used in this run was 8388608.
==794== 
==794== HEAP SUMMARY:
==794==     in use at exit: 1,130 bytes in 12 blocks
==794==   total heap usage: 15 allocs, 3 frees, 10,346 bytes allocated
==794== 
==794== 4 bytes in 1 blocks are still reachable in loss record 1 of 7
==794==    at 0x4849D8C: malloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x48E35FF: strdup (strdup.c:42)
==794==    by 0x108C3F: agregar_puffle (in /root/volume/probar_setup/program)
==794==    by 0x10A6EF: main (in /root/volume/probar_setup/program)
==794== 
==794== 16 bytes in 1 blocks are still reachable in loss record 2 of 7
==794==    at 0x484C0A4: calloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x109E87: main (in /root/volume/probar_setup/program)
==794== 
==794== 22 bytes in 4 blocks are still reachable in loss record 3 of 7
==794==    at 0x4849D8C: malloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x48E35FF: strdup (strdup.c:42)
==794==    by 0x108C3F: agregar_puffle (in /root/volume/probar_setup/program)
==794==    by 0x10A067: main (in /root/volume/probar_setup/program)
==794== 
==794== 48 bytes in 2 blocks are still reachable in loss record 4 of 7
==794==    at 0x484C0A4: calloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x109EB3: main (in /root/volume/probar_setup/program)
==794== 
==794== 96 bytes in 2 blocks are still reachable in loss record 5 of 7
==794==    at 0x484C0A4: calloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x108DBF: agregar_penguin (in /root/volume/probar_setup/program)
==794==    by 0x109FCF: main (in /root/volume/probar_setup/program)
==794== 
==794== 472 bytes in 1 blocks are still reachable in loss record 6 of 7
==794==    at 0x4849D8C: malloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x48C9C97: __fopen_internal (iofopen.c:65)
==794==    by 0x109D9F: main (in /root/volume/probar_setup/program)
==794== 
==794== 472 bytes in 1 blocks are still reachable in loss record 7 of 7
==794==    at 0x4849D8C: malloc (in /usr/lib/aarch64-linux-gnu/valgrind/vgpreload_memcheck-arm64-linux.so)
==794==    by 0x48C9C97: __fopen_internal (iofopen.c:65)
==794==    by 0x109DBF: main (in /root/volume/probar_setup/program)
==794== 
==794== LEAK SUMMARY:
==794==    definitely lost: 0 bytes in 0 blocks
==794==    indirectly lost: 0 bytes in 0 blocks
==794==      possibly lost: 0 bytes in 0 blocks
==794==    still reachable: 1,130 bytes in 12 blocks
==794==         suppressed: 0 bytes in 0 blocks
==794== 
==794== For lists of detected and suppressed errors, rerun with: -s
==794== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
Segmentation fault
root@9806622cfb78:~/volume/probar_setup#

**Reply by 4n1t4** [reply]
Consulta en cómo estas corriendo tu programa?
No estas utilizando devcontainer o similar?

**Comment by SantiagoOsorio3** [comment]
Edd-docker

Obtener Outlook para iOS<https://aka.ms/o0ukef>
________________________________
De: 4n1t4 ***@***.***>
Enviado: Tuesday, August 20, 2024 11:29:54 AM
Para: IIC2133-PUC/2024-2 ***@***.***>
Cc: Santiago Osorio Holmes ***@***.***>; Author ***@***.***>
Asunto: Re: [IIC2133-PUC/2024-2] AYUDA EN ERROR DE MEMORIA EN LEAVE PENGUIN (Discussion #41)


Consulta en cómo estas corriendo tu programa?
No estas utilizando devcontainer o similar?

**Reply by 4n1t4** [reply]
Dale dale ya vi tu código base, hay un par de cositas q igual le hectárea un ojo pero si prefieres podemos hablar por discord, puedo el bloque después de almuerzo, sino , sería después de las 5. Te tinca?

**Reply by SantiagoOsorio3** [reply]
Feliz, eso si podria desde las 6.. mi telegram es @santiagoosorio3 para que nos pongamos de acuerdo desde ahi

**Reply by SantiagoOsorio3** [reply]
perdon, es en minuscula, se cambio autaticamente por el de guthub.
